script refactoring

klkvr opened the pull request:
## Motivation

The idea is to refactor script to be more like a state-machine, so its easier to navigate/update and we are able to debug each state transition separately.

## Solution

I've started with `build` module, it now returns and operates on `BuildData` and `LinkedBuildData` structs which contain all build artifacts and solc-related data and allow us to use them through script execution instead of passing `Linker` everywhere.

Closes #7244 
Closes #4476

mattsse reviewed (commented):
I like this approach a lot.

structuring this in compile+link steps makes a lot of sense

klkvr commented:
Refactoring goals were mostly increasing readability and safety of the code.

Readability right now should be better because the logic overall became more flat. If earlier we had `handle_broadcastable_transactions` going up to 4-5 functions deep, it's now split across several states which are invoked sequentially and don't go deeper than 2 small functions.

Safety isn't great because we have a lot of `Option`s which are _assumed_ to be `Some` after some point. It is not really easy to know where this point is and it does not help both reading existing code and extending it. I've tried to reduce such practicies as well. Example of those is ScriptConfig`: https://github.com/foundry-rs/foundry/blob/c24933da985419ea143de7e8636d5b0a48d2fab7/crates/forge/bin/cmd/script/mod.rs#L619-L625

All of those parameters are initialized with default values and populated later, accessing those in the wrong place might result in panic or unintended behavior.

I've also simplified logic where it was possible, however, we still have several pretty long functions which would be nice in the shorter form, but this can be fixed a bit later, esp considering that all of the state transaction functions are normally readable

klkvr commented:
We should also be able to remove onchain simulation step once we have isolation mode enabled by default

mattsse commented on crates/forge/bin/cmd/script/cmd.rs:
```diff
@@ -1,376 +1,99 @@
-                &predeploy_libraries,
-                script_wallets.clone(),
-            )
+        let pre_simulation = self
+            .preprocess()
+            .await?
+            .compile()?
+            .link()?
+            .prepare_execution()
+            .await?
+            .execute()
+            .await?
+            .prepare_simulation()
             .await?;
```
this is great!

I'd like a few more function docs that cover the workflow

mattsse reviewed (changes requested):
this looks great, this is a lot easier to navigate

I'd like to see a few more docs on functions,types,fields

mattsse reviewed (approved):
the individual states are very easy to follow now.

@DaniPopes would appreciate a closer look

DaniPopes commented on crates/forge/bin/cmd/script/sequence.rs:
```diff
@@ -16,16 +16,91 @@ use foundry_common::{
+        }
+    }
+
+    pub fn iter_sequences(&self) -> impl Iterator<Item = &ScriptSequence> {
```
`fn sequences{_mut}(&self) -> &[ScriptSequence]` rather than iter IMO

DaniPopes commented on crates/forge/bin/cmd/script/mod.rs:
```diff
@@ -1,64 +1,59 @@
-use alloy_json_abi::{Function, InternalType, JsonAbi};
-use alloy_primitives::{Address, Bytes, Log, U256, U64};
-use alloy_rpc_types::request::TransactionRequest;
+
```
no spaces in between imports

DaniPopes reviewed (changes requested):
Can we extract this into a separate crate?

mattsse commented:
> Can we extract this into a separate crate?

oh yeah, totally forgot about this, I'd like that, yes

klkvr commented:
@DaniPopes @mattsse I fthink it makes sense to avoid having `clap` dependency in the separate crate. A reasonable way to do this might be to move some of `ScriptArgs` methods to `ScriptConfig` and populate them in `ScriptArgs::preprocess`. But we'd duplicate half of `ScriptArgs` fields that way.

Do you see any cleaner solution here?

klkvr commented:
ah, we'd still need clap for `VerifyArgs`, so we can either line up verification extraction and get rid of clap where it will be possible then, or divide `ScriptArgs` and `ScriptConfig` now anyway

klkvr commented:
Extracted scripting code to separate crate and also removed `ScriptArgs` dependency on `BuildArgs`. It now uses `CoreBuildArgs` + separate `--skip` parameter. Removed params are `--names`, `--sizes`, `--watch*` and `--format-json`. All of those were not used and had not effect. This change should close #4476

mattsse commented on crates/script/src/build.rs:
```diff
@@ -0,0 +1,231 @@
+        // Otherwise, parse input as <path>:<name> and use the path from the contract info, if
+        // present.
+        let target_path = if let Ok(path) = dunce::canonicalize(&args.path) {
+            Ok::<_, eyre::Report>(Some(path))
+        } else {
+            let contract = ContractInfo::from_str(&args.path)?;
+            target_name = Some(contract.name.clone());
+            if let Some(path) = contract.path {
+                Ok(Some(dunce::canonicalize(path)?))
+            } else {
+                Ok(None)
+            }
+        }?;
```
is the additional Ok wrapper required here?
I dont think we have an err, hence the `Ok<_, eyre>`

mattsse commented on crates/script/src/cmd.rs:
```diff
@@ -0,0 +1,111 @@
+use foundry_cli::utils::LoadConfig;
+use foundry_common::{shell, types::ToAlloy};
+
+impl ScriptArgs {
```
this should be below scriptargs type def

mattsse commented on crates/script/src/execute.rs:
```diff
@@ -0,0 +1,493 @@
+    pub abi: JsonAbi,
+}
+
+impl LinkedState {
```
this should be below type def

klkvr replied:
yeah, I've placed it like this as we had it split across several files before, should I get rid of `states.rs` at all and keep impls of all states along with their definitions?

mattsse replied:
that'd be great, I prefer that style because then you don't need to jump across multiple files when looking at a type:

```rust
struct A

impl A

impl Trait for A

```

although @DaniPopes prefers `impl Trait` above the `impl A` block, I'm indifferent about the order of these two

mattsse reviewed (changes requested):
amazing.

I see that a few `impl` blocks are not colocated with thir struct def, this is probably because the previous ScriptArgs impl was scattered across multiple files, but this makes it a lot harder to navigate.

impl blocks should go under their respective struct def

mattsse reviewed (approved):
lgtm

pending @DaniPopes

DaniPopes commented on crates/script/src/sequence.rs:
```diff
@@ -26,6 +23,86 @@ use std::{
+        }
+    }
+
+    pub fn get_sequence_mut(&mut self, index: usize) -> Option<&mut ScriptSequence> {
```
I think all these methods are redundant if we can just use `fn sequences() -> &[ScriptSequence]`

klkvr replied:
The need for `get_sequence_mut` and `sequences_len` is from the `BundledState::broadcast` where we need mutable access to separate sequences to populate pending txes and receipts, but we also sometimes need mutable access to the parent `ScriptSequenceKind` to do checkpoint save of all sequences to the file. so not sure what's the best approach to reduce redundancy here

DaniPopes replied:
right I mean this
```rust
    pub fn sequences(&self) -> &[ScriptSequence] {
        match self {
            ScriptSequenceKind::Single(sequence) => std::slice::from_ref(sequence),
            ScriptSequenceKind::Multi(sequence) => &sequence.deployments,
        }
    }

    pub fn sequences_mut(&mut self) -> &mut [ScriptSequence] {
        match self {
            ScriptSequenceKind::Single(sequence) => std::slice::from_mut(sequence),
            ScriptSequenceKind::Multi(sequence) => sequence.deployments,
        }
    }
```

Then all the other functions (`get_sequence_mut`, `sequences_len`) can just call the corresponding methods on the slices returned by these methods

klkvr replied:
ah got it, updated in 5bd705d3ca951b6b0828fa7e4e74ed58c90b9814

DaniPopes commented on crates/script/src/sequence.rs:
```diff
@@ -26,6 +23,86 @@ use std::{
+
+impl Drop for ScriptSequenceKind {
+    fn drop(&mut self) {
+        self.save(false, true).expect("could not save deployment sequence");
```
Should probably be `if let Err(e) = { error!() }` rather than panicking on IO errors

DaniPopes commented on crates/script/src/multi_sequence.rs:
```diff
@@ -0,0 +1,152 @@
+    pub deployments: Vec<SensitiveScriptSequence>,
+}
+
+fn to_sensitive(sequence: &mut MultiChainSequence) -> SensitiveMultiChainSequence {
```
Does this need `&mut`? Also I'd prefer this be like `SensitiveMultiChainSequence::from_multi_sequence` or a From impl

DaniPopes commented on crates/script/src/broadcast.rs:
```diff
@@ -0,0 +1,429 @@
+            .collect::<Vec<_>>();
+
+        let errors =
+            join_all(futs).await.into_iter().filter(|res| res.is_err()).collect::<Vec<_>>();
```
we can `filter_map(Result::err)` and format the list using itertools `eyre!("{}", errors.iter().format("\n"))`
